### ISSUE TITLE
union type appears to be incorrectly merged, resulting in false-positive InvalidReturnStatement

### ISSUE THREAD

**Issue by WDD-Marv** [vimeo/psalm#948]
when using a method to conditionally return either an array of types, or the count of the types (i.e. an object repository), with the return type being docblocked as `array<int, T>|int`, psalm instead complains of the return type being `array<int, T|int>`

https://getpsalm.org/r/4e7f10dc72

>>> 

> Psalm output (using commit cb63f4f): 
>
> ERROR: InvalidReturnStatement - 9:5 - The type 'array<int, string|int>' does not match the declared return type 'array<int, string>' for Foo::ObtainArray
> 
> ERROR: InvalidReturnType - 5:5 - The declared return type 'array<int, string>' for Foo::ObtainArray is incorrect, got 'array<int, string|int>'

```php
<?php
class Foo
{
  /**
  * @return array<int, string>
  */
  public function ObtainArray() : array
  {
    return (array) $this->ObtainCountOrArray(true);
  }
  
  public function ObtainCount() : int
  {
    return (int) $this->ObtainCountOrArray(false);
  }
  
  /**
  * @return array<int, string>|int
  */
  protected function ObtainCountOrArray(bool $arrayNotCount)
  {
    return $arrayNotCount ? [1 => 'one'] : 0;
  }
}
```

**Comment by muglug** [comment]
because
```php
(array) (rand(0, 1) ? $a : $b)
```
is the same as 
```php
rand(0, 1) ? (array) $a : (array) $b
```

**Comment by WDD-Marv** [comment]
but
```php
rand(0, 1) ? [] : 1
```
is not the same as 
```php
(array) (rand(0, 1) ? [] : 1)
```

**Comment by muglug** [comment]
Indeed.

But you have `return (array) $this->ObtainCountOrArray(true);` where `ObtainCountOrArray` returns `array<int, string>|int`.

I think you want Psalm to think that `(array)` means "ignore the int part", but that's not how Psalm works.

**Comment by WDD-Marv** [comment]
yeah, that's what I was treating it as.

this is a little more readable
```php
return (array) $this->doThing();
```

than this:
```php
/**
* @var array<int, string> $out
*/
$out = $this->doThing();

return $out;
```
i.e. https://getpsalm.org/r/5339acc536

is there a verbose logging option for psalm that would indicate when psalm is coercing docblock types? it would be handy to know when I'm either doing something funky or my docblock is wrong :)

**Comment by weirdan** [comment]
Would be solved easily if Psalm supported #698 (overloading)
